### ISSUE TITLE
Allow scanning of files that are not in namespaces but consume namespaces

### ISSUE THREAD

**Issue by AshleyPinner** [icanhazstring/composer-unused#11]
Currently the code seems to only scan psr0/4, classmap and files directives. However, we have a large Cake2 project where the bulk of our files are not namespaced. They still use composer dependencies.

An example high level directory structure is like (in our case):
```
bin/
Config/
Console/
Controller/
Lib/
Model/
Plugin/
src/
Test/
vendor/
View/
webroot/
composer.json
composer.lock
```

While `/src` contains a section of our code which is namespaced, our MVC files are in either the Controller/Model/View directories, or in the Plugin directory. These have no namespace, but do consume namespaces.

It would be handy if the tool could scan those files as well, as currently the output of unused is correct _if we are only talking about the src directory_. The rest of the "unused" packages are consumed in our MCV structure (or at least I'd like to think so!)

**Comment by icanhazstring** [comment]
Hi @AshleyPinner thanks for reporting.

This is indeed something I will add in the future.

Some questions tho:
1. As you listed `vendor/` as well, does this also contain non composer packages?
2. Can't you provide these folders inside your `classmap` directive in your `composer.json`

**Comment by AshleyPinner** [comment]
1. Yes, but they are put into the classmap, so from the point of "is this namespace in use" it should be the same detection method
2. I'm unable to put our MCV code into the classmap, and wouldn't want to for many reasons (such as duplicate file names and the fact that the MCV code totally lacks any namespace declarations, and adding them breaks many things)

Granted, this is a slightly more obscure usage, but I suspect there's a lot of code out there that follows the pattern of a vendor directory with composer packages in it, and code that includes the autoloader and rolls forward from that point, without a namespace.

**Comment by icanhazstring** [comment]
Ok I will see if I can come up with some sort of additional paths to scan :)

**Comment by TomKriek** [comment]
For a Lumen project we have a file in the bootstrap directory `app.php`.

The first few lines look like this.

```
<?php

require_once __DIR__.'/../vendor/autoload.php';

try {
    (new Dotenv\Dotenv(__DIR__.'/../'))->load();
} catch (Dotenv\Exception\InvalidPathException $e) {
    //
}
```

But the vlucas/phpdotenv package is being marked as unused. Having extra config to include/exclude paths would be amazing.

**Comment by xuanquynh** [comment]
Could you mark this issue as "enhancement"?

**Comment by icanhazstring** [comment]
We could add an additional cli parameter which uses an array of folders to be scanned as well.
Something like this:

```
$ composer unused --add=folderA/ --add=folderB/
```

**Comment by nicholasruunu** [comment]
@icanhazstring I believe that an `--exclude` parameter combined with glob for `**/*.php` instead would be a better option. Although, perhaps too slow?

**Comment by icanhazstring** [comment]
> @icanhazstring I believe that an `--exclude` parameter combined with glob for `**/*.php` instead would be a better option. Although, perhaps too slow?

Yep. I already hit a peak with a company project about 81K files. Which crashed my VM 😄 
But on the other hand - yours is a good choice too.

Maybe something like `--force-complete` which does a complete scan of every folder beside your `composer.json` so then we can use an `--exclude` to ignore smt like `vendor/` and stuff.

**Comment by AshleyPinner** [comment]
That works. As long as you can avoid the circular reference of scanning vendor 😄 (maybe vendor should be auto-excluded?)

**Comment by icanhazstring** [comment]
Yep good point :D

**Comment by icanhazstring** [comment]
Feature idea:
Add two new cli arguments:
- `--force-complete` which scan all directories beside `composer.json` (using glob `**/*.php`)
- `--exclude=` array of arguments to exclude from scanning
- auto exclude `vendor/`

**Comment by AshleyPinner** [comment]
> auto exclude vendor/

Sort of obvious but please ensure you read the config.vendor-dir key, as vendor directories are not always /vendor

(Our project just lives to be an exception to many rules)

**Comment by nicholasruunu** [comment]
> Feature idea:
> Add two new cli arguments:
> 
> * `--force-complete` which scan all directories beside `composer.json` (using glob `**/*.php`)
> * `--exclude=` array of arguments to exclude from scanning
> * auto exclude `vendor/`

If not extremely slow I'd think that `--force-complete` should just be the default and overwritten with `--directories` or something.

I think it will be extremely common to just run this as default and remove deps which might not be obviously used in the project. For example script files that are run by cron. I think the less dangerous path to follow is to check all your php-files.

**Comment by icanhazstring** [comment]
> I think it will be extremely common to just run this as default and remove deps which might not be obviously used in the project. For example script files that are run by cron. I think the less dangerous path to follow is to check all your php-files.

Good point. 👍

**Comment by icanhazstring** [comment]
I just reworked the usage scan to use all `*.php` files in the project without `vendor/`
If you install `dev-master` you can give it a try.

I will add another cli argument for excluding folders from the console.

**Comment by icanhazstring** [comment]
Just released version `0.3.0` to address this kind of issue.
If the problem persists please let me know.

Closing it for now :)
